src/main/java/frc/robot/scoreStates/ShootState.java

```diff
@@ -44,20 +60,49 @@ public Command driveToClosestOptimalPoint() {
                     return AutoBuilder.pathfindToPose(
                             new Pose2d(
                                     optimalTranslation,
-                                    Utils.calculateOptimalRotation(
+                                    Utils.calcRotationToTranslation(
                                             optimalTranslation, speakerPose)),
                             Constants.AUTO_CONSTRAINTS);
                 },
                 Set.of(SwerveDrive.getInstance()));
     }
 
     @Override
-    public Command stateInitialize() {
-        return null;
+    public Command initializeCommand() {
+        return Commands.none();
+    }
+
+    @Override
+    public Command initializeSubsystem() {
+        return Commands.parallel(
+                shooter.setVelocity(
+                        () ->
+                                Units.RotationsPerSecond.of(
+                                                ShooterConstants.interpolationMap.get(
+                                                                Utils.getDistanceFromPoint(
+                                                                        speakerPose,
+                                                                        PoseEstimation.getInstance()
+                                                                                .getEstimatedPose()))
+                                                        .value)
+                                        .mutableCopy()),
```

Comment:
If you have one statement that ranges _10 lines_, you should probably write some variables. Note that the `Utils.getDistanceFromPoint(...)` is the same between both calculations.

Reply:
We rewrote this code in 60aa413

Reply:
Well, now you have this code duplicated four times.
About the in-bounds, is it going to be addressed in this branch or Emma's? Anyway, this is a solved problem in geometry, look for "point in poligon".

Reply:
I switched some of the statement to a constant.
I think calculating whether the point is in the polygon or not isn't necessary. The current calculation isn't very long.

Reply:
This is not relevant we fixed this in 60aa413
It is not possible for the robot to shoot through the stage anymore

Reply:
And what about the "don't warm up when we're far"? Where is this going to be eventually addressed? If in another branch, please open an issue for it.

Reply:
a16a75cbdbdab75b2a6e06c44abe20d7e945b135

Reply:
1. `shooterToSpeakerHeight` should be determined at runtime.
2. Why not use `inBounds`?
3. Why are there different warm-up distances?
4. Please use regular function blocks, long statements hinder readability.
5. Is `prepareSubsystems()` going to be run once or continuously? I raised some concerns about the former here: https://github.com/Galaxia5987/robot-2024/pull/39#discussion_r1485198415

Reply:
1. 1f98b9da4215075b473ecf153d4b4aa148f09a10
2. You want to warmup a bit before you're in bounds.
3. They're not really distances from the speaker but a line a bit before the wing line, I'll change the name to a more fitting one. 29a037688875eb7824da3ac009fb9185c7b16c87
5. 6cc1d01f85f60d2f771ef020168dbf5bc9f501da
6. It's going to run once, I addressed those concerns.

Reply:
1. 👍🏻
2. 👍🏻
3. My question is why do you need blue and red ones and not just a single constant.
4. 👍🏻
5. Where?
6. While it's not necessary to change now, I believe real testing might show that the angle to set the hood won't be just the arctangent of the angle (for example in longer distances the Note's trajectory will curve more and might require a higher angle). So just prepare yourselves for the option that an angle interpolation map will be needed.

Reply:
3. They're different for blue and red.
5. https://github.com/Galaxia5987/robot-2024/pull/39#discussion_r1485203194
6. 👍

Reply:
OK, I understand the usage now, but:
1. The values don't make sense to me -- why is `MAX_BLUE_WARMUP_LINE` so close to the red alliance's wall?
2. Why not just check the distance to the relevant Speaker?

And two more things:
1. Use [`Commands.onlyIf()`](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/Command.html#onlyIf(java.util.function.BooleanSupplier)) instead of `either()` with an empty command.
2. Why not run `setHood()` along with `setShooter()` on line 100? There is no reason to move the hood if the shooter doesn't run.

Reply:
1. The line just an random value or now
2. Checking the distance is just overkill and not actually required
3. 09a9d9d
4. The setHood happens parallel to the setShooter

Reply:
I don't think it's overkill, on the contrary, I think it's simpler because you'll have only one constant and you won't need to have both `<` and `>` in the condition.

About the hood, my suggestion is that the `onlyIf()` will apply to the hood, too.

Reply:
1. We have better things to work on and it barely matters, Plus we don't care that it will activate one second before it is required.
2. You are contradicting yourself please see the other shoot state issue, Resolve the redundant one and continue in the one that you didn't resolve

Reply:
> we don't care that it will activate one second before it is required.

This is not related to my suggestion. All I'm saying is that instead of having two constants for two lines, you could have one constant for the warm-up distance. Then your condition will be shorter because you'll only do `distance from my alliance's Speaker to robot < warm-up distance`. At the very least, you can make `MAX_BLUE_WARMUP_LINE` be `width of the field - MAX_RED_WARMUP_LINE`. That way you'll have only one constant to change.
> You are contradicting yourself please see the other shoot state issue

I don't. My suggestion is that the function will look like this:
```diff
diff --git a/src/main/java/frc/robot/scoreStates/ShootState.java b/src/main/java/frc/robot/scoreStates/ShootState.java
index 191cc22..ca89260 100644
--- a/src/main/java/frc/robot/scoreStates/ShootState.java
+++ b/src/main/java/frc/robot/scoreStates/ShootState.java
@@ -95,8 +95,7 @@ public class ShootState implements ScoreState {

     @Override
     public Command prepareSubsystems() {
-        return Commands.parallel(
-                setShooter()
+        return setShooter().alongWith(setHood())
                         .onlyIf(
                                 () ->
                                         (isRed()
@@ -106,9 +105,8 @@ public class ShootState implements ScoreState {
                                                 || ((!isRed()
                                                         && botPose.getX()
                                                                 > ShooterConstants
-                                                                        .MAX_BLUE_WARMUP_LINE))),
-                setHood(),
-                CommandGroups.getInstance().retractGrillevator());
+                                                                        .MAX_BLUE_WARMUP_LINE)))
+                                .alongWith(CommandGroups.getInstance().retractGrillevator());
     }

     @Override
```
Here it is in non-`git diff` view:
```java
    @Override
    public Command prepareSubsystems() {
        return setShooter()
                .alongWith(setHood())
                .onlyIf(
                        () ->
                                (isRed() && botPose.getX() < ShooterConstants.MAX_RED_WARMUP_LINE)
                                        || ((!isRed()
                                                && botPose.getX()
                                                        > ShooterConstants.MAX_BLUE_WARMUP_LINE)))
                .alongWith(CommandGroups.getInstance().retractGrillevator());
    }
```
That way the hood too will warm up only when needed. I hope that clears things up. Remember you call also call me if you think that will help.

Reply:
I don't think it matters all that much but its a bit more readable.
468b3600c570bf58bc35577fe5b3cfe0cf7bb023

Reply:
> I don't think it matters all that much

Yep. But small optimizations like this one add up and in the end the code is much better.
What about my suggestions on the warm-up distance(s)? It's ok if you don't want to implement them, but I'd like to hear your opinion on them, whatever that is.

Reply:
a14442e

Reply:
You don't need code for both alliances now...

Reply:
0957f385f48d6c1c2a74d348ddfa412fbeefa31e

Reply:
`Translation2d.getDistance()` :-)

Reply:
But I don't want the direct distance from the bot to the speaker, I want the difference between the x values.

Reply:
I'm not sure I understand why -- the purpose is to give the shooter enough time to warm up, so we should measure time (= distance), not entering some area or not. But I want to merge this so let's talk about it when I come to school in a few days